### PR TITLE
Ensure free buffer space when reading TLS messages

### LINKED ISSUE #83455: SslStream fails to authenticate in .NET 7.0 but works in .NET 6.0
### Description

I have an app targeting .NET 6.0, which uses a `SslStream` to establish a TLS connection to a remote server and transfer data through it. After I updated the app to .NET 7.0, the `AuthenticateAsClientAsync()` call started to throw an exception in some cases:

```StackTrace
Unhandled exception. System.AggregateException: One or more errors occurred. ( Received an unexpected EOF or 0 bytes from the transport stream.)
 ---> System.IO.IOException:  Received an unexpected EOF or 0 bytes from the transport stream.
   at System.Net.Security.SslStream.EnsureFullTlsFrameAsync[TIOAdapter](CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.PoolingAsyncValueTaskMethodBuilder`1.StateMachineBox`1.System.Threading.Tasks.Sources.IValueTaskSource<TResult>.GetResult(Int16 token)
   at System.Net.Security.SslStream.ReceiveBlobAsync[TIOAdapter](CancellationToken cancellationToken)
   at System.Net.Security.SslStream.ForceAuthenticationAsync[TIOAdapter](Boolean receiveFirst, Byte[] reAuthenticationData, CancellationToken cancellationToken)
```

Nothing else is changed in the app, except the TargetFramework value - with .NET 6.0 it works, with .NET 7.0 it fails.

I also tested the app on other OS and there were NO exceptions there (Windows 11 22H2, macOS Ventura, Ubuntu 22)

### Reproduction Steps

Execute the following code from a console app targeting .NET 7.0 on Windows 10:

```C#
using System.Net;
using System.Net.Security;
using System.Net.Sockets;
using System.Security.Authentication;
using System.Security.Cryptography.X509Certificates;


var targetHost = "lh3.googleusercontent.com";
var ip = Dns.Resolve(targetHost); // resolves to 142.250.186.129
Socket baseSocket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
baseSocket.Connect(new IPEndPoint(ip.AddressList[0], 443));
Stream strmNet = new NetworkStream(baseSocket, false);
var httpsStream = new SslStream(strmNet, false);
var appProtocols = new List<SslApplicationProtocol>() { SslApplicationProtocol.Http2, SslApplicationProtocol.Http11 };
var oAcceptedProtocols = SslProtocols.Tls | SslProtocols.Tls11 | SslProtocols.Tls12;
var opt = new SslClientAuthenticationOptions()
{
    TargetHost = targetHost,
    ClientCertificates = null,
    EnabledSslProtocols = oAcceptedProtocols,
    CertificateRevocationCheckMode = X509RevocationMode.NoCheck,
    ApplicationProtocols = appProtocols
};

var ct = new CancellationToken(false);
httpsStream.AuthenticateAsClientAsync(opt, ct).Wait();
Console.WriteLine("Stream is authenticated? " + httpsStream.IsAuthenticated.ToString());
```

### Expected behavior

AuthenticateAsClientAsync should not throw an exception when trying to authenticate.

### Actual behavior

The code throws an exception:

Unhandled exception. System.AggregateException: One or more errors occurred. ( Received an unexpected EOF or 0 bytes from the transport stream.)

### Regression?

Yes, the code worked fine on .NET 6.0

### Known Workarounds

No workaround at the moment :(

### Configuration

Failing on:
OS version: Windows 10 x64 22H2
.NET version: 7.0.201

Works on
OS version: Windows 11 x64 22H2, macOS Ventura, Ubuntu 22.04
.NET version: 6.0.100 and 7.0.201

### Other information

_No response_

### PR DESCRIPTION
**Pull request by rzikm**
Fixes #83455.

### CODE REVIEW

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs [hunk 1]
```diff
@@ -715,15 +715,10 @@ private async ValueTask<int> EnsureFullTlsFrameAsync<TIOAdapter>(CancellationTok
...
-            {
-                _buffer.EnsureAvailableSpace(frameSize - _buffer.EncryptedLength);
-            }
-
             while (_buffer.EncryptedLength < frameSize)
             {
-                // there should be space left to read into
-                Debug.Assert(_buffer.AvailableLength > 0, "_buffer.AvailableBytes > 0");
+                // make sure we have space to read into
+                _buffer.EnsureAvailableSpace(Math.Min(frameSize, _buffer.Capacity) - _buffer.EncryptedLength);
```
Comment (wfurt) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs:721]: Is there risk of growing too much if framers are keep coming. I know this is not normal case but as far as I understand the old code, we would make space for a frame, maybe more if needed. 
We had `EnsureAvailableSpace` on line 246 so I'm wondering why that was not sufficient.

**Comment by wfurt** [comment]
I assume test would be difficult?

File: src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs (hunk 1 above)
Comment (rzikm) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs:721]: yes, there is, I should've waited with the fix till the morning :D

> We had EnsureAvailableSpace on line 246 so I'm wondering why that was not sufficient.

Not sure which one you are talking about, the closes one is in 214 and 287 and both are outside of the while loop which is receiving TLS frames

**Comment by rzikm** [comment]
> I assume test would be difficult?

yes, the exact condition for this regression is that there are just enough leftover bytes at the end of the buffer that we don't know the size of the next frame (when debugging it was 3 or 4 bytes). With more bytes, the frame size would be known and the check on 720 would fix that.

I am not even sure we can make a test for this, since we can't control the size of TLS frames in handshake in order to hit the condition above.

**Comment by rzikm** [comment]
Test failures are
- https://github.com/dotnet/runtime/issues/83481
- https://github.com/dotnet/runtime/issues/83482

**Review by wfurt — APPROVED** [review]
LGTM.

**Pull request merged**